### PR TITLE
FIX-#2134: Fix mismatch partitioning insertions with same index

### LINKED ISSUE #2134: `__setitem__` after `read_csv` with a `Series` from pandas causes metadata error
### Describe the problem

`__setitem__` after `read_csv` with a `Series` from pandas causes metadata error

### Source code / logs

```python
In [1]: import modin.pandas as pd                                                                                                                                              
df ^[[A
In [2]: df = pd.read_csv("test_1mx256.csv")                                                                                                                                    

In [3]: df["new"] = pd.Series(list(range(len(df))))                                                                                                                            
```

Error
```
UserWarning: Distributing <class 'list'> object. This may take some time.

In [4]: df                                                                                                                                                                     
Out[4]: ---------------------------------------------------------------------------
Exception                                 Traceback (most recent call last)
/Library/Frameworks/Python.framework/Versions/3.6/lib/python3.6/site-packages/IPython/core/formatters.py in __call__(self, obj)
    700                 type_pprinters=self.type_printers,
    701                 deferred_pprinters=self.deferred_printers)
--> 702             printer.pretty(obj)
    703             printer.flush()
    704             return stream.getvalue()

/Library/Frameworks/Python.framework/Versions/3.6/lib/python3.6/site-packages/IPython/lib/pretty.py in pretty(self, obj)
    400                         if cls is not object \
    401                                 and callable(cls.__dict__.get('__repr__')):
--> 402                             return _repr_pprint(obj, self, cycle)
    403 
    404             return _default_pprint(obj, self, cycle)

/Library/Frameworks/Python.framework/Versions/3.6/lib/python3.6/site-packages/IPython/lib/pretty.py in _repr_pprint(obj, p, cycle)
    695     """A pprint that just redirects to the normal repr function."""
    696     # Find newlines and replace them with p.break_()
--> 697     output = repr(obj)
    698     for idx,output_line in enumerate(output.splitlines()):
    699         if idx:

~/software_builds/modin/modin/pandas/dataframe.py in __repr__(self)
    168 
    169             num_cols += len(self.columns) - i
--> 170         result = repr(self._build_repr_df(num_rows, num_cols))
    171         if len(self.index) > num_rows or len(self.columns) > num_cols:
    172             # The split here is so that we don't repr pandas row lengths.

~/software_builds/modin/modin/pandas/base.py in _build_repr_df(self, num_rows, num_cols)
    117         else:
    118             indexer = row_indexer
--> 119         return self.iloc[indexer]._query_compiler.to_pandas()
    120 
    121     def _update_inplace(self, new_query_compiler):

~/software_builds/modin/modin/backends/pandas/query_compiler.py in to_pandas(self)
    193 
    194     def to_pandas(self):
--> 195         return self._modin_frame.to_pandas()
    196 
    197     @classmethod

~/software_builds/modin/modin/engines/base/frame/data.py in to_pandas(self)
   1816             ErrorMessage.catch_bugs_and_request_email(
   1817                 not df.index.equals(self.index) or not df.columns.equals(self.columns),
-> 1818                 "Internal and external indices do not match.",
   1819             )
   1820             df.index = self.index

~/software_builds/modin/modin/error_message.py in catch_bugs_and_request_email(cls, failure_condition, extra_log)
     52                 "Internal Error. "
     53                 "Please email bug_reports@modin.org with the traceback and command that"
---> 54                 " caused this error.\n{}".format(extra_log)
     55             )
     56 

Exception: Internal Error. Please email bug_reports@modin.org with the traceback and command that caused this error.
Internal and external indices do not match.
```

### PR DESCRIPTION
**Pull request by devin-petersohn**
Signed-off-by: Devin Petersohn <devin.petersohn@gmail.com>

## What do these changes do?

Add a simple shuffle internal mechanism to have more performant insertions and concatenation when index objects already match.

- [x] commit message follows format outlined [here](https://modin.readthedocs.io/en/latest/contributing.html)
- [x] passes `flake8 modin`
- [x] passes `black --check modin`
- [x] signed commit with `git commit -s` 
- [x] Resolves #2134  
- [x] tests added and passing

### CODE REVIEW

File: modin/engines/base/frame/partition_manager.py [hunk 1]
```diff
@@ -280,6 +280,50 @@ def map_axis_partitions(cls, axis, partitions, map_func, keep_partitioning=False
         # the structure to the correct order.
         return result_blocks.T if not axis else result_blocks
 
+    @classmethod
+    def shuffle_simple(cls, axis, partitions, map_func, lengths):
```
Comment (aregm) [inline modin/engines/base/frame/partition_manager.py:284]: Isn't it better to name simple_shuffle to match the data.py change?

**Review by aregm — APPROVED** [review]

File: modin/engines/base/frame/data.py [hunk 2]
```diff
@@ -1618,6 +1623,36 @@ def _copartition(self, axis, other, how, sort, force_repartition=False):
             reindexed_other_list.append(reindexed_other)
         return reindexed_self, reindexed_other_list, joined_index
 
+    def _simple_shuffle(self, axis, other):
+        """
+        Shuffle other rows or columns to match partitioning of self.
+
+        Note: This should only be called when `other.axes[axis]` and `self.axes[axis]`
```
Comment (YarShev) [inline modin/engines/base/frame/data.py:1630]:
```suggestion
        Notes
        -----
        This should only be called when `other.axes[axis]` and `self.axes[axis]`
```

Comment (YarShev) [reply modin/engines/base/frame/data.py:1630]: `Notes` should be follow after `Returns`.

File: modin/engines/base/frame/data.py [hunk 3]
```diff
@@ -1618,6 +1623,36 @@ def _copartition(self, axis, other, how, sort, force_repartition=False):
...
+            are identical.
+
+        Parameters
+        ----------
+            axis: 0 or 1
+                The axis to shuffle over
+            other: BasePandasFrame
+                The BasePandasFrame to match to this object's partitioning
```
Comment (YarShev) [inline modin/engines/base/frame/data.py:1638]:
```suggestion
                The BasePandasFrame to match to self object's partitioning
```

File: modin/pandas/test/dataframe/test_indexing.py [hunk 4]
```diff
@@ -1156,6 +1156,16 @@ def test___setitem__(data):
     df_equals(modin_df, pandas_df)
 
 
+def test___setitem__with_mismatched_partitions():
+    fname = "200kx99.csv"
+    np.savetxt(fname, np.random.randint(0, 100, size=(200_000, 99)), delimiter=",")
+    modin_df = pd.read_csv(fname)
+    pandas_df = pandas.read_csv(fname)
+    modin_df["new"] = pd.Series(list(range(len(modin_df))))
+    pandas_df["new"] = pd.Series(list(range(len(pandas_df))))
```
Comment (YarShev) [inline modin/pandas/test/dataframe/test_indexing.py:1165]:
```suggestion
    pandas_df["new"] = pandas.Series(list(range(len(pandas_df))))
```

File: modin/engines/base/frame/partition_manager.py [hunk 5]
```diff
@@ -280,6 +280,50 @@ def map_axis_partitions(cls, axis, partitions, map_func, keep_partitioning=False
...
+            lengths : list(int)
+                The list of lengths to shuffle the object
+
+        Returns
+        -------
+        NumPy array
+            An array of new partitions for a Modin Frame.
+        """
+        preprocessed_map_func = cls.preprocess_func(map_func)
+        partitions = (
```
Comment (YarShev) [inline modin/engines/base/frame/partition_manager.py:305]: There is the function [axis_partition](https://github.com/modin-project/modin/blob/fe0afedd051f9aa721d27a63fcb8c67df32d9768/modin/engines/base/frame/partition_manager.py#L78) to get `column_partitions` or `row_partitions`. You need to merge with master before use this.

**Review by YarShev — CHANGES_REQUESTED** [review]

**Comment by YarShev** [comment]
@devin-petersohn , I am okay with the changes but CI is red.

**Comment by devin-petersohn** [comment]
@YarShev it looks like a Ray core dump in I/O which shouldn't have been touched by this. I restarted CI to see if it magically goes away.

**Comment by devin-petersohn** [comment]
CI is red here looks like due to Omnisci. @ienkovich take a look?

**Comment by Garra1980** [comment]
And probably @dchigarev

**Comment by dchigarev** [comment]
Conda failed to download packages because of a server error, I've re-run CI to see if it help

**Comment by dchigarev** [comment]
@devin-petersohn @YarShev CI is green now

**Review by YarShev — APPROVED** [review]

**Pull request merged**